runner/data.go

```diff
@@ -44,6 +44,16 @@ type StreamMessageProviderFunc func(*CallData) (*dynamic.Message, error)
 // Clients can return ErrEndStream to end the call early
 type StreamRecvMsgInterceptFunc func(*dynamic.Message, error) error
 
+// StreamInterceptorProviderFunc is an interface for a function invoked to generate a stream interceptor
+type StreamInterceptorProviderFunc func() StreamInterceptor
+
+// StreamInterceptor is an interface for sending and receiving stream messages.
+// The interceptor can keep shared state for the send and receive calls.
+type StreamInterceptor interface {
+	Recv(*dynamic.Message, error) error
+	Send(*CallData) (*dynamic.Message, error)
```

Comment:
What's in `CallData`?

One thought is, maybe it should be an input of the inteceptor func instead
```go
type StreamInterceptorProviderFunc func(*CallData) StreamInterceptor

type StreamInterceptor interface {
	Recv(*dynamic.Message, error) error
	Send() (*dynamic.Message, error)
}
```

Reply:
It could be, but the nice thing about defining Send this way is that it has the same signature as a `StreamMessageProviderFunc`

Reply:
CallData has a mix of the method level info and some request/response level specific info.
It's not very well designed for streaming RPCs.

You can keep it in `Send()`.
But I think the `StreamInterceptorProviderFunc` will also need it (if the interceptor's behavior is different for different RPCs).

Reply:
Ok makes sense